expand.go

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"bytes"
+	"golang.org/x/text/transform"
+	"io"
+)
+
+// expandReadCloser implements io.ReadCloser
+type expandReadCloser struct {
+	tReader *transform.Reader
+	closer  io.Closer
+}
+
+func newExpandReadCloser(rc io.ReadCloser, expandFunc func(string) (string, error)) *expandReadCloser {
+	return &expandReadCloser{
+		tReader: transform.NewReader(
+			rc,
+			&expandTransformer{expand: expandFunc}),
+		closer: rc,
+	}
+}
+
+func (this *expandReadCloser) Read(p []byte) (int, error) {
```

Comment:
I consider that advice from the go project to be the worst kind of advice.  "Use single letter variable names".  Garbage.  When you see the letter 'b' in a function, should you immediately assume that it refers to the calling object?  What about the letter 'c'?  I often use the letter 'c' for a variable that holds a single character.  However, if you see a variable named "this", every developer will immediately understand that it refers to the calling object.

Now, I understand that "methods" in Go are much more similar to the way you do objects in C, by passing in the object as the first argument.  So while a Go method definition might look like this:

   func (*ObjectType) foo (...)

it is equivalent to a function defined like this:

   func foo (*ObjectType, ...) 

So what?  It does not improve readability or remove ambiguity by using a variable named 'o' to refer to the object which called this function.  Why Rob Pike thinks this is a good idea, and actively rails against using something that I consider both more readable and less ambiguous, is beyond me.

Python has the same calling convention for methods as Go does, i.e. the calling object is passed in as the first argument and can be given any name by the developer.  At least that project had the sense to understand that there is zero value in using a different name for the calling object in every different class, and there is a negative impact on readability, so they suggest calling it "self".  That's a convention worth documenting IMHO.

But, it's your project so I'll use whatever letter you want me to use.  I can't think of a single letter that would be immediately recognized by a random developer as referring to the calling object.  Please tell me which letter you'd like me to use.